library/pkg/executor/executor.go

```
@@ -0,0 +1,359 @@
+//
+// in case there is any problem with the setup of the process, these cgroup controls cannot be created an error will be returned
+// In case the process setup was successful a Command will be returned
+// The Command will either have "running" or "failed" state based on whether the process was succesfully started
+// Further interaction with the process should happen through the returned Command instance
+func (e Executor) StartJob(processId string, command string, args []string) (Command, error) {
+	err := e.createCgroup(processId)
+
+	if err != nil {
+		cleanCgroups(processId)
+		return nil, err
+	}
+
+	cmd := exec.Command(command, args...)
+
+	outputBuffer := &internal.MultiplexingBuffer{}
+
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = outputBuffer
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Cloneflags: syscall.CLONE_NEWUTS | syscall.CLONE_NEWPID | syscall.CLONE_NEWUSER | syscall.CLONE_NEWNET | syscall.CLONE_NEWNS,
+		UidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getuid(),
+				Size:        1,
+			},
+		},
+		GidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getgid(),
+				Size:        1,
+			},
+		},
+	}
+
+	err = cmd.Start()
+
+	if err != nil {
+		return &internal.FailedCommand{}, nil
+	}
+
+	pid := cmd.Process.Pid
+
+	err = e.addPidToCgroups(processId, pid)
+	if err != nil {
+		killErr := cmd.Process.Kill()
+
+		if killErr != nil {
+			return nil, fmt.Errorf("the process could not be added to cgroup and killed in cleanup %w", killErr)
+		}
+
+		return nil, fmt.Errorf("failed to add process to cgroup %w", killErr)
+	}
+
+	c := &internal.StartedCommand{
+		Cmd:          cmd,
+		CmdStatus:    internal.RUNNING,
+		OutputBuffer: outputBuffer,
+		AfterDone: func() {
+			cleanCgroups(processId)
+		},
+	}
+	return c, nil
+}
+
+func cleanCgroups(processId string) {
+	processCgroupFolder := filepath.Join(cgroupFolder, processId)
+
+	_ = syscall.Rmdir(processCgroupFolder)
```

ignored error.

---

my thinking was the following: this is a cleanup step, there is really no good way to propagate this back to the caller, and I am not even sure that this should be interesting for the caller. It's totally internal to the executor lib to create and cleanup cgroups, the caller has no clue about how and when that happens. 

So barring returning an error and somehow propagating that back to the caller, i have two options, logging, or panicing. Neither is something I really want a library to do.

The third thing I can do now, is to propagate a state "CLEANUP_FAILED" to the caller, but that would hide the actual status of the process, therefore I am a bit hesitant. Also returning "finished", error("failed to cleanup") would be a bit strange. Therefore I decided to ignore this error, there is really not much more we can do with it